### ISSUE TITLE
Conditional animations

### ISSUE THREAD

**Issue by Tolriq** [bumptech/glide#511]
Is there a simple integrated way to handle the load animation based on a condition not at the call time but at the loading time ?

Typical basic usage : `.animate(getUserVisibleHint() ? R.anim.abc_fade_in : R.anim.none)` with a viewpager that preload fragments.

This works well when loading from cache as it's fast so 99% of the time the image is loaded in invisible fragment before displaying. But when loading from Internet it can be slow and so the image can be loaded while the fragment is visible. 

And of course in that case the animation does not play since the call was made when the fragment was not visible.

This would also maybe allow a proper handling of animate / don't animate in more global way avoiding the empty anim to cover this, since you can't pass 0 or -1 to animate to disable or a boolean somewhere. Making the usage of the builder pattern less useful.

**Comment by TWiStErRob** [comment]
There's a `.dontAnimate()` on the builder, apart from that you can handle the animation yourself at many points (I used lambda methods to prevent clutter):

##### in a `.listener()`
The animation is not even constructed if not needed, even though it's provided at call time.
```java
.animate(R.anim.abc_fade_in)
.listener(onResourceReady -> {
	if (!getUserVisibleHint()) {
		target.onResourceReady(resource, null);
		return true; // will skip default behavior
		// @see GenericRequest.onResourceReady(resource, result)
	}
	return false;
})
```

##### encapsulate in a custom target
Animation is constructed in GenericRequest, but thrown away.
```java
.into(new GlideDrawableImageViewTarget(image) {
	onResourceReady ->
			super.onResourceReady(resource, getUserVisibleHint()? animation : null);
});
```

##### via `.animate(AnimationFactory)`
I would like this solution best, but currently this version of animate has default visibility.
@sjudd, why is that method not `public`?
```java
.animate(new ViewAnimationFactory<GlideDrawable>(context, R.anim.abc_fade_in) {
	build -> {
		if (!getUserVisibleHint()) {
			return NoAnimation.get();
		} else {
			return super.build(isFromMemoryCache, isFirstResource);
		}
	}
})
```
The visibility problem can be avoided with a little hack:
```java
package com.bumptech.glide;
public class BuilderAccessor {
	public static <ModelType> DrawableRequestBuilder<ModelType>
	animate(DrawableRequestBuilder<ModelType> builder, ViewAnimationFactory<GlideDrawable> factory) {
		builder.animate(factory);
		return builder;
	}
}
```

##### Global
How do you think a global way would work? You need the Fragment/Activity to decide whether to animate or not to animate.

**Comment by Tolriq** [comment]
I refrained for putting a disclaimer for you about answering too fast I should have ;)

Anyway thanks for the listener thing if it works, but show me how to use conditional animation with animate and dontanimate in a single query ?

Either needs duplicate code or split the builder calls with an if in the middle in both case not a great way to use a builder.

Passing a bool to dontanimate could help for example, but builders with on / off commands in different function name are not good for builder pattern.

**Comment by TWiStErRob** [comment]
> you about answering too fast

I'm pretty sure you would get the same or similar answer from @sjudd, maybe less wordy :)

> show me how to use conditional animation with animate and dontanimate in a single query?

I thought your original problem was that deciding too early doesn't work sometimes. If you go with a solution that always works you can just pass in an animation and it will be skipped later anyway, so no need for tricky conditionals in the builder. But yes, if you go with the 99% it's not really possible with a single line. Your solution is a good workaround though. *Btw, are you sure the animation is executed if the view is not even laid out?*

> builders with on / off commands in different function name are not good for builder pattern

I thought the Builder pattern was only meant to replace huge constructors (mostly because Java doesn't support optional args), and/or make more meaningful names for primitive setters or group multiple setters together to avoid unknown dependencies. If you check Wikipedia for example, the sample there has `set` and `unset` methods much like `[dont]Animate`. Nothing really suggest that it should be used as a one-statement build, thought that's how most are implemented for convenience. So in the end I agree, I don't like splitting my builders either. I had a similar problem (needed an `if` around the builder) with `.load(null)`, that's the reason Sam added the `.fallback()` method in 268.

> Passing a bool to dontanimate could help

`.dontAnimate(false)` doesn't really make sense for me, because then how do you tell which anim to use? If `dontAnimate` is a toggle, meaning it's sole purpose is to enable and disable the animation it may be confusing for most people.
**But**, I would be pro `animate(0)` and `animate((Animator)null)` being the same as `dontAnimate()`, it's compatible and logical and is easy to implement. @sjudd, do you think that's feasible?

**Comment by Tolriq** [comment]
You did answer about the builder problem with the .dontanimate leading to my answer ...

And yes the animations are played, I would not bother to do that if there was no performance problem with all animations running on large images.

For the builder, the main problem is that Glide builder is not a real builder by it's own, but a multitude of  builders with type change depending on order of the commands.

So having for example a generic function that create the first part of the query then complement it, you have to deal with incompatible type and duplicate codes and complexity.

(Like BitmapRequestBuilder<ImageRequest, Bitmap>  and DrawableRequestBuilder<ImageRequest>) So not having needs to cut the builder pattern in multiple part is nearly mandatory due to this.

For a normal builder pattern cutting the build and adding if in the middle is less beautiful but not really problematic.

**Comment by TWiStErRob** [comment]
Thanks for your thoughts, I like when I learn something from you! I think these issues will be addressed in 4.0 with Options, though I didn't look deep into the source yet.

**Comment by sjudd** [comment]
I believe this is possible with animation factorys and transition options in Glide v4. Let me know if you think something is still missing from those APIs.
